test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestFeatureService.java

```diff
@@ -10,57 +10,136 @@
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
-import java.util.NavigableMap;
+import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
-import java.util.function.Predicate;
+import java.util.function.BiConsumer;
+
+import static java.util.Collections.emptySet;
 
 class ESRestTestFeatureService implements TestFeatureService {
-    private final Predicate<String> historicalFeaturesPredicate;
     private final Set<String> clusterStateFeatures;
     private final Set<String> allSupportedFeatures;
+    private final Set<String> knownHistoricalFeatureNames;
 
-    ESRestTestFeatureService(
-        List<? extends FeatureSpecification> specs,
-        Collection<Version> nodeVersions,
-        Set<String> clusterStateFeatures
-    ) {
-        var minNodeVersion = nodeVersions.stream().min(Comparator.naturalOrder());
-        var featureData = FeatureData.createFromSpecifications(specs);
-        var historicalFeatures = featureData.getHistoricalFeatures();
-        Set<String> allHistoricalFeatures = historicalFeatures.lastEntry() == null ? Set.of() : historicalFeatures.lastEntry().getValue();
-
-        this.allSupportedFeatures = Sets.union(clusterStateFeatures, minNodeVersion.<Set<String>>map(v -> {
-            var historicalFeaturesForVersion = historicalFeatures.floorEntry(v);
-            return historicalFeaturesForVersion == null ? Set.of() : historicalFeaturesForVersion.getValue();
-        }).orElse(allHistoricalFeatures));
-
-        this.historicalFeaturesPredicate = minNodeVersion.<Predicate<String>>map(
-            v -> featureId -> hasHistoricalFeature(historicalFeatures, v, featureId)
-        ).orElse(featureId -> true); // We can safely assume that new non-semantic versions (serverless) support all historical features
+    ESRestTestFeatureService(List<FeatureSpecification> featureSpecs, Collection<Version> nodeVersions, Set<String> clusterStateFeatures) {
+        List<FeatureSpecification> specs = new ArrayList<>(featureSpecs);
+        specs.add(new RestTestLegacyFeatures());
+        if (MetadataHolder.HISTORICAL_FEATURES != null) {
+            specs.add(MetadataHolder.HISTORICAL_FEATURES);
+        }
+        var historicalFeatures = FeatureData.createFromSpecifications(specs).getHistoricalFeatures();
+        this.knownHistoricalFeatureNames = Optional.ofNullable(historicalFeatures.lastEntry()).map(Map.Entry::getValue).orElse(emptySet());
         this.clusterStateFeatures = clusterStateFeatures;
+        this.allSupportedFeatures = Sets.union(
+            clusterStateFeatures,
+            nodeVersions.stream()
+                .min(Comparator.naturalOrder())
+                .map(v -> Optional.ofNullable(historicalFeatures.floorEntry(v)).map(Map.Entry::getValue).orElse(emptySet()))
+                .orElse(knownHistoricalFeatureNames)
```

Comment:
This is a bit confusing - having an Optional inside an Optional.map call. When is emptySet used vs knownHistoricalFeatureNames? Can this be a bit more explicit here and less stream-y?

Reply:
With above change, the internal optional / emptySet fallback is obsolete.
Thanks for pointing this out, this is much easier to read know 👍